R/join_keys.R

```diff
@@ -0,0 +1,673 @@
+  }
+  x <- rlang::list2(...)
+  UseMethod("join_keys", x[[1]])
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.join_keys <- function(...) {
+  if (length(...) > 1) {
+    return(join_keys.default(...))
+  }
+  x <- rlang::list2(...)
+  x[[1]]
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.teal_data <- function(...) {
+  if (length(...) > 1) {
+    return(join_keys.default(...))
+  }
+  x <- rlang::list2(...)
+  x[[1]]@join_keys
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.TealData <- function(...) {
+  x <- rlang::list2(...)
+  if (length(...) > 1) {
+    return(join_keys.default(...))
+  }
+  x[[1]]$get_join_keys()
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.default <- function(...) {
+  x <- rlang::list2(...)
+
+  # Constructor
+  res <- new_join_keys()
+  if (length(x) > 0) {
+    join_keys(res) <- x
+  }
+
+  res
+}
+
+#' @rdname join_keys
+#' @details
+#' The setter assignment `join_keys(obj) <- ...` will merge obj and `...` if obj
+#' is not empty.
+#'
+#' @param join_keys_obj (`join_keys`) empty object to set the new relationship pairs.
+#' @param value (`join_key_set` or list of `join_key_set`) relationship pairs to add
+#' to `join_keys` list.
+#'
+#' @export
+`join_keys<-` <- function(join_keys_obj, value) {
+  UseMethod("join_keys<-", join_keys_obj)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Using the setter (assignment) ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' join_keys(jk) <- join_key("ds3", "ds4", "some_col2")
+#' join_keys(jk)["ds1", "ds3"] <- "some_col3"
+#' jk
+`join_keys<-.join_keys` <- function(join_keys_obj, value) {
+  if (missing(value)) {
+    return(join_keys_obj)
+  }
+
+  # Assume assignment of join keys as a merge operation
+  #  Needed to support join_keys(jk)[ds1, ds2] <- "key"
+  if (checkmate::test_class(value, classes = c("join_keys", "list"))) {
+    return(merge_join_keys(join_keys_obj, value))
+  }
+
+  # Assignment of list of join_key_set will merge it with existing join_keys
+  if (length(join_keys_obj) > 0 && checkmate::test_list(value, types = "join_key_set", min.len = 1)) {
+    jk <- new_join_keys()
+    join_keys(jk) <- value
+    return(merge_join_keys(join_keys_obj, jk))
+  }
+
+  if (inherits(value, "join_key_set")) value <- list(value)
+
+  checkmate::assert_list(value, types = "join_key_set", min.len = 1)
+
+  # check if any join_key_sets share the same datasets but different values
+  for (idx_1 in seq_along(value)) {
+    for (idx_2 in seq_along(value[idx_1])) {
+      assert_compatible_keys(value[[idx_1]], value[[idx_2]])
+    }
+    join_keys_obj <- join_pair(join_keys_obj, value[[idx_1]])
+  }
+
+  logger::log_trace("join_keys keys are set.")
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#'
+#' @examples
+#'
+#' c(join_keys(join_key("a", "b", "c")), join_keys(join_key("a", "d2", "c")))
+c.join_keys <- function(...) {
+  x <- rlang::list2(...)
+
+  if (!length(x)) {
+    return(NULL)
+  }
```

Comment:
Will this ever occur? Empty `...` will not be dispatched to this method, will it?

Reply:
Probably not, but there is always a chance of it happend if called directly `c.join_keys()`

Other examples keep this protection _(example on `c.difftime`)_

```r
function (..., recursive = FALSE) 
{
  coerceTimeUnit <- function(x) {
    switch(attr(x, "units"), secs = x, mins = 60 * x, hours = 60 * 
      60 * x, days = 60 * 60 * 24 * x, weeks = 60 * 60 * 
      24 * 7 * x)
  }
  args <- list(...)
  if (!length(args)) 
    return(.difftime(double(), "secs"))
# ...
```

Reply:
True, but not all:
```
getS3method("c", "POSIXct")()
Error in class(xx) <- cl : attempt to set an attribute on NULL
```

I'm not going to press it because it's a minor detail but I believe method dispatch provides enough checks, if someone calls a specific method the responsibility is on them.